starknet_py/net/models/abi/definition_test.py

```diff
@@ -0,0 +1,334 @@
+    "delete_pool", OrderedDict(id=pool_id_struct, user_id=uint256_struct), OrderedDict()
+)
+
+
+def test_parsing_types_abi():
+    # Even though user depend on pool id and uint256 it is defined first. Parser has to consider those cases
+    abi = AbiDefinition.from_list(
+        [
+            user_dict,
+            pool_id_dict,
+            uint256_dict,
+            user_added_dict,
+            pool_id_added_dict,
+            get_user_dict,
+            delete_pool_dict,
+        ]
+    )
+
+    assert abi.defined_structures == {
+        "Uint256": uint256_struct,
+        "PoolId": pool_id_struct,
+        "User": user_struct,
+    }
+    assert abi.events == {
+        "UserAdded": user_added_event,
+        "PoolIdAdded": pool_id_added_event,
+    }
+    assert abi.functions == {"get_user": get_user_fn, "delete_pool": delete_pool_fn}
+
+
+def test_duplicated_structure():
+    with pytest.raises(
+        AbiDecodingError,
+        match="Name 'Uint256' was used more than once in defined structures",
+    ):
+        AbiDefinition.from_list([uint256_dict, pool_id_dict, uint256_dict])
+
+
+def test_duplicated_function():
+    with pytest.raises(
+        AbiDecodingError,
+        match="Name 'get_user' was used more than once in defined functions",
+    ):
+        AbiDefinition.from_list(
+            [get_user_dict, delete_pool_dict, get_user_dict, delete_pool_dict]
+        )
+
+
+def test_duplicated_event():
+    with pytest.raises(
+        AbiDecodingError,
+        match="Name 'UserAdded' was used more than once in defined events",
+    ):
+        AbiDefinition.from_list([user_added_dict, delete_pool_dict, user_added_dict])
+
+
+def test_duplicated_type_members():
+    type_dict = {
+        "type": "struct",
+        "name": "Record",
+        "size": 4,
+        "members": [
+            {"name": "name", "offset": 0, "type": "felt"},
+            {"name": "value", "offset": 1, "type": "felt"},
+            {"name": "id", "offset": 2, "type": "felt"},
+            {"name": "value", "offset": 3, "type": "felt"},
+        ],
+    }
+    with pytest.raises(
+        AbiDecodingError,
+        match="Name 'value' was used more than once in members of structure 'Record'",
+    ):
+        AbiDefinition.from_list([type_dict])
+
+
+def test_missing_type_used_by_other_type():
+    with pytest.raises(UnknownCairoTypeError, match="Type 'Uint256' is not defined"):
+        AbiDefinition.from_list([pool_id_dict])
+
+
+def test_missing_type_used_by_function():
+    with pytest.raises(UnknownCairoTypeError, match="Type 'Uint256' is not defined"):
+        AbiDefinition.from_list([get_user_dict])
+
+
+def test_missing_type_used_by_event():
+    with pytest.raises(UnknownCairoTypeError, match="Type 'User' is not defined"):
+        AbiDefinition.from_list([user_added_dict])
+
+
+def test_deserialize_proxy_abi():
+    # Contains all types of ABI apart from structures
+    abi = json.loads(read_contract("oz_proxy_abi.json"))
```

Comment:
We should already have a fixture for this contract that could be used here.

Reply:
We only use its abi, not contract itself.